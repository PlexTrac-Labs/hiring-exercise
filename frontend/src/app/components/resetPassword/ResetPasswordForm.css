.ResetPasswordFrame {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-top: 200px;
}

.ResetPasswordTitle {
  display: flex;
  flex-direction: column;
  padding: 10px;
  font-size: 2em;
}

.ResetPasswordForm {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px;
}

.ResetPasswordInput {
  padding-bottom: 10px;
}

.ResetPasswordButtonBox {
  display: flex;
  flex-direction: row;
  justify-content: center;
  padding-top: 40px;
}

.ResetPasswordButton {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.ResetPasswordButtonSpacer {
  min-width: 10px;
}

.ErrorMessage {
  display: flex;
  flex-direction: column;
  align-items: center;
  color: #920000;
  font-weight: bold;
  padding-top: 10px;
  padding-bottom: 10px;
}
