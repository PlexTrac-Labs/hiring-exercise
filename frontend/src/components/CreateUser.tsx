import { ListItem, ListItemButton, ListItemIcon } from "@mui/material";
import React from "react";
import { LIST_MIN_WIDTH } from "../helpful/constants";
import AddCircleIcon from "@mui/icons-material/AddCircle";

export const CreateUser = () => {
  const style = {
    minWidth: LIST_MIN_WIDTH,
  };

  return (
    <ListItem sx={style}>
      <ListItemButton>
        <ListItemIcon>
          <AddCircleIcon color="success" />
        </ListItemIcon>
        Create User
      </ListItemButton>
    </ListItem>
  );
};
