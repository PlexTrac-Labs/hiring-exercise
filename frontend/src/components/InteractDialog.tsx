import { Dialog, DialogContent, DialogTitle } from "@mui/material";
import React from "react";
import { colors } from "../helpful/style";

export type InteractDialogProps = {
  open: boolean;
  title: string;
  handleClose: () => void;
};

export const InteractDialog: React.FC<InteractDialogProps> = (props) => {
  const { children, open, title, handleClose } = { ...props };

  const TitleStyle = {
    backgroundColor: colors.darkGreen,
    color: "white",
  };

  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle sx={TitleStyle}>{title}</DialogTitle>
      <DialogContent>{children}</DialogContent>
    </Dialog>
  );
};
