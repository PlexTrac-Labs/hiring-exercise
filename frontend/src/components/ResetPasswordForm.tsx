import { Alert, Button, FormControl, Grid, TextField } from "@mui/material";
import React from "react";

export type ResetFormProps = {
  error: boolean;
  errorMessage: string;
  handleSubmit: (event: any) => void;
};

const defaultProps: Partial<ResetFormProps> = {};

export const ResetPasswordForm = (props: ResetFormProps) => {
  const { error, errorMessage, handleSubmit } = {
    ...defaultProps,
    ...props,
  };
  return (
    <form onSubmit={handleSubmit}>
      <Grid
        container
        alignItems="center"
        justifyContent="center"
        rowSpacing={2}
        marginTop={3}
      >
        {error && (
          <Grid item xs={12}>
            <Alert severity="error">{errorMessage}</Alert>
          </Grid>
        )}
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="old"
              variant="outlined"
              label="Current Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="new"
              variant="outlined"
              label="New Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="confirm"
              variant="outlined"
              label="Confirm New Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12}>
          <Button color="primary" variant="contained" type="submit">
            Change Password
          </Button>
        </Grid>
      </Grid>
    </form>
  );
};
