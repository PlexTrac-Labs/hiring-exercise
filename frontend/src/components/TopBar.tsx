import { AppBar, Button, Grid, Toolbar } from "@mui/material";
import { Box, styled } from "@mui/system";
import React from "react";
import { useNavigate } from "react-router-dom";
import { AUTH_TOKEN, TOP_BAR_HEIGHT } from "../helpful/constants";
import { colors } from "../helpful/style";

export const TopBar = () => {
  const loggedIn = !!localStorage.getItem(AUTH_TOKEN);
  const navigate = useNavigate();

  const Bar = styled(AppBar)(() => ({
    height: TOP_BAR_HEIGHT,
    backgroundColor: colors.darkGreen,
  }));

  const handleLogout = () => {
    localStorage.removeItem(AUTH_TOKEN);
    navigate("/");
  };

  return (
    <Bar>
      <Toolbar>
        {loggedIn && (
          <Grid
            container
            direction="row"
            justifyContent="space-between"
            alignItems="center"
          >
            <Grid item>
              <Button color="inherit" onClick={() => navigate("/list")}>
                Users
              </Button>
              <Button color="inherit" onClick={() => navigate("/reset")}>
                Reset Password
              </Button>
            </Grid>
            <Button color="inherit" onClick={handleLogout}>
              Logout
            </Button>
          </Grid>
        )}
      </Toolbar>
    </Bar>
  );
};
