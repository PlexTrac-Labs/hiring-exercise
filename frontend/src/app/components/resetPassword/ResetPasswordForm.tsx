import { FormEvent, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button, TextField } from '@material-ui/core';
import { useStore } from '../../stores/store';
import { ResetPasswordFormValues } from '../../models/resetPasswordFormValues';
import BasicModal, { BasicModalProps } from '../modal/BasicModal';
import { useStyles } from '../../globalStyles';
import './ResetPasswordForm.css';

const ResetPasswordForm = () => {
  const classes = useStyles();
  const navigate = useNavigate();

  const {
    sessionStore: { resetPassword }
  } = useStore();

  const [showModal, setShowModal] = useState(false);
  const [email, setEmail] = useState('');
  const [currentPassword, setCurrentPassword] = useState('');
  const [newPassword, setNewPassword] = useState('');
  const [confirmNewPassword, setConfirmNewPassword] = useState('');
  const [errorMessage, setErrorMessage] = useState('');

  let resetModalProps: BasicModalProps = {
    bodyText: 'Password reset successfully.',
    showOkButton: true,
    showCancelButton: false,
    handleOk: () => handleCancel(),
    handleCancel: () => {}
  };

  const handleReset = (event: FormEvent) => {
    event.preventDefault();
    if (!validateForm()) {
      return;
    }

    if (showModal) {
      return;
    }

    const formValues: ResetPasswordFormValues = {
      email,
      currentPassword,
      newPassword
    };

    resetPassword(formValues)
      .then(res => {
        setShowModal(true);
      })
      .catch(error =>
        setErrorMessage('Error processing your reset password request')
      );
  };

  const handleCancel = () => {
    navigate('/navigator', { replace: true });
  };

  const handleEmailChange = (e: any) => {
    setEmail(e.target.value);
  };

  const handleCurrentPasswordChange = (e: any) => {
    setCurrentPassword(e.target.value);
  };

  const handleNewPasswordChange = (e: any) => {
    setNewPassword(e.target.value);
  };

  const handleConfirmNewPasswordChange = (e: any) => {
    setConfirmNewPassword(e.target.value);
  };

  const validateForm = (): boolean => {
    setErrorMessage('');

    if (email === '') {
      setErrorMessage('Enter your email address');
      return false;
    }

    if (currentPassword === '') {
      setErrorMessage('Enter your current password');
      return false;
    }

    if (newPassword === '') {
      setErrorMessage('Enter your new password');
      return false;
    }

    if (confirmNewPassword === '') {
      setErrorMessage('Enter confirmation password');
      return false;
    }

    if (newPassword !== confirmNewPassword) {
      setErrorMessage('Passwords do not match');
      return false;
    }

    if (newPassword.length < 6) {
      setErrorMessage('Password must be at least 6 characters');
      return false;
    }

    if (currentPassword === newPassword) {
      setErrorMessage('New password must be different');
      return false;
    }

    return true;
  };

  const hasFormErrors = (): boolean => {
    return errorMessage !== '';
  };

  return (
    <div className='ResetPasswordFrame'>
      <div className='ResetPasswordTitle'>Reset Password</div>
      <div className='ResetPasswordForm'>
        <form
          className={classes.root}
          noValidate
          autoComplete='off'
          onSubmit={handleReset}
        >
          <div className='ResetPasswordInput'>
            <TextField
              name='email'
              id='email'
              placeholder='email'
              value={email}
              onChange={handleEmailChange}
              autoFocus
            />
          </div>
          <div className='ResetPasswordInput'>
            <TextField
              name='currentPassword'
              id='currentPassword'
              placeholder='current password'
              type='password'
              value={currentPassword}
              onChange={handleCurrentPasswordChange}
            />
          </div>
          <div className='ResetPasswordInput'>
            <TextField
              name='password'
              id='password'
              placeholder='new password'
              type='password'
              value={newPassword}
              onChange={handleNewPasswordChange}
            />
          </div>
          <div className='ResetPasswordInput'>
            <TextField
              name='confirmPassword'
              id='confirmPassword'
              placeholder='confirm new password'
              type='password'
              value={confirmNewPassword}
              onChange={handleConfirmNewPasswordChange}
            />
          </div>
          {hasFormErrors() && (
            <div className='ErrorMessage'>{errorMessage}</div>
          )}
          <div className='ResetPasswordButtonBox'>
            <div className='ResetPasswordButton'>
              <Button variant='contained' color='primary' type='submit'>
                Reset
              </Button>
            </div>
            <div className='ResetPasswordButtonSpacer'></div>
            <div className='ResetPasswordButton'>
              <Button variant='contained' type='button' onClick={handleCancel}>
                Cancel
              </Button>
            </div>
          </div>
        </form>
      </div>
      <div>
        {showModal && (
          <BasicModal {...resetModalProps} header={'Password Reset'} />
        )}
      </div>
    </div>
  );
};

export default ResetPasswordForm;
