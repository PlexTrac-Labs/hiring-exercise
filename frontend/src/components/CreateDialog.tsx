import React, { useState } from "react";
import { CreateUserRequest, User } from "../util/user/types";
import { createUser, getAllUsers } from "../util/user/user";
import { InteractDialog } from "./InteractDialog";
import { UserForm } from "./UserForm";

export type CreateDialogProps = {
  open: boolean;
  setUsers: (users: User[]) => void;
  onClose: () => void;
};

export const CreateDialog = (props: CreateDialogProps) => {
  const { open, setUsers, onClose } = { ...props };
  const [error, setError] = useState(false);
  const errorMessage = "Problem creating user. Is this email already in use?";

  const handleClose = () => {
    setError(false);
    onClose();
  };

  const handleCreate = async (event: any) => {
    event.preventDefault();
    const payload: CreateUserRequest = {
      username: event.target.username.value,
      password: event.target.password.value,
      firstName: event.target.firstName.value,
      lastName: event.target.lastName.value,
      email: event.target.email.value,
      birthYear: parseInt(event.target.birthYear.value),
      favColor: event.target.favColor.value,
    };
    try {
      await createUser(payload);
      let users = await getAllUsers();
      setUsers([...users]);
      handleClose();
    } catch (error: any) {
      setError(true);
    }
  };

  return (
    <InteractDialog open={open} handleClose={handleClose} title="Create User">
      <UserForm
        showPassword
        submitText="Create"
        error={error}
        errorMessage={errorMessage}
        handleSubmit={handleCreate}
        handleClose={handleClose}
      />
    </InteractDialog>
  );
};
