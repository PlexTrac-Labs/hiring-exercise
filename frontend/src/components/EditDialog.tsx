import React, { useState } from "react";
import { UpdateUserRequest, User } from "../util/user/types";
import { getAllUsers, updateUser } from "../util/user/user";
import { InteractDialog } from "./InteractDialog";
import { UserForm } from "./UserForm";

export type EditDialogProps = {
  open: boolean;
  user: User;
  setUsers: (users: User[]) => void;
  onClose: () => void;
};

export const EditDialog = (props: EditDialogProps) => {
  const { open, user, setUsers, onClose } = { ...props };
  const [error, setError] = useState(false);
  const errorMessage = "Problem updating user. Is this email already in use?";

  const handleClose = () => {
    setError(false);
    onClose();
  };

  const handleUpdate = async (event: any) => {
    event.preventDefault();
    const payload: UpdateUserRequest = {
      username: event.target.username.value,
      firstName: event.target.firstName.value,
      lastName: event.target.lastName.value,
      email: event.target.email.value,
      birthYear: parseInt(event.target.birthYear.value),
      favColor: event.target.favColor.value,
    };
    try {
      await updateUser(user._id, payload);
      let users = await getAllUsers();
      setUsers([...users]);
      handleClose();
    } catch (error: any) {
      setError(true);
    }
  };

  return (
    <InteractDialog open={open} handleClose={handleClose} title="Edit User">
      <UserForm
        user={user}
        submitText="Save"
        error={error}
        errorMessage={errorMessage}
        handleSubmit={handleUpdate}
        handleClose={handleClose}
      />
    </InteractDialog>
  );
};
