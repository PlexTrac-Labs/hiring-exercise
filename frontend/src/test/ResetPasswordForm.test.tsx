import { cleanup, render } from "@testing-library/react";
import React from "react";
import { ResetPasswordForm } from "../components/ResetPasswordForm";

afterEach(cleanup);

it("renders", async () => {
  const { asFragment } = render(<ResetPasswordForm />);
  expect(asFragment()).toMatchSnapshot();
});
