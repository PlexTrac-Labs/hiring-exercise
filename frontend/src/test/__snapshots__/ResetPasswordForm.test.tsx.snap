// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column css-1irzm6x-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="old"
              id="old-label"
            >
              Current Password
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="old"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Current Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="new"
              id="new-label"
            >
              New Password
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="new"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    New Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="confirm"
              id="confirm-label"
            >
              Confirm New Password
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="confirm"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Confirm New Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Change Password
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
